### PR TITLE
fsck: Add the capability to fix zero size directory

### PR DESCRIPTION
**Pull request by Erichorng**
Add the ability to fix the directory which size=0 and starts with 0 cluster. This kind of directory often made by Paragon ufsd dirver, and it will cause failure when we try to it using linux exfat driver. The way to fix it, we should find a cluster and assign it to the directory's first cluster and modify the size to 1 (Like exfat driver does during mkdir).

### CODE REVIEW

File: fsck/fsck.c [hunk 1]
```diff
@@ -215,6 +226,39 @@ static int check_clus_chain(struct exfat_de_iter *de_iter,
 	}
 
 	return 0;
+locate_cluster:
+	node->size = count * exfat->clus_size;
```
Comment (hclee) [inline fsck/fsck.c:230]: Why this statement is needed?

File: fsck/fsck.c [hunk 2]
```diff
@@ -215,6 +226,39 @@ static int check_clus_chain(struct exfat_de_iter *de_iter,
...
+	    (ssize_t)exfat->clus_size) {
+		exfat_err("failed to fill new cluster with zeroes\n");
+		return -EIO;
+	}
+	exfat_info("final cluster : %#x \n", new_clus);
+	count = 1;
+	stream_de->stream_start_clu = cpu_to_le64(new_clus);
+	stream_de->stream_size = cpu_to_le64(count * exfat->clus_size);
+	stream_de->stream_valid_size = cpu_to_le64(count * exfat->clus_size);
+	stream_de->dentry.stream.flags = 3;
```
Comment (hclee) [inline fsck/fsck.c:256]: Could you use macro constant instead of magic value?

File: fsck/fsck.c [hunk 3]
```diff
@@ -215,6 +226,39 @@ static int check_clus_chain(struct exfat_de_iter *de_iter,
...
+	/* zero out the new cluster */
+	if (exfat_write(exfat->blk_dev->dev_fd, exfat->zero_cluster,
+			exfat->clus_size, exfat_c2o(exfat, new_clus)) !=
+	    (ssize_t)exfat->clus_size) {
+		exfat_err("failed to fill new cluster with zeroes\n");
+		return -EIO;
+	}
+	exfat_info("final cluster : %#x \n", new_clus);
+	count = 1;
+	stream_de->stream_start_clu = cpu_to_le64(new_clus);
```
Comment (hclee) [inline fsck/fsck.c:253]: cpu_to_le32() is correct.

File: fsck/repair.c [hunk 4]
```diff
@@ -61,6 +61,7 @@ static struct exfat_repair_problem problems[] = {
 	{ER_FILE_LARGER_SIZE, ERF_PREEN_YES, ERP_TRUNCATE, 0, 0, 0},
 	{ER_FILE_DUPLICATED_CLUS, ERF_PREEN_YES, ERP_TRUNCATE, 0, 0, 0},
 	{ER_FILE_ZERO_NOFAT, ERF_PREEN_YES, ERP_FIX, 0, 0, 0},
+	{ER_DIR_FIRST_CLUS, ERF_PREEN_YES, ERP_TRUNCATE, 0, 0, 0}
```
Comment (hclee) [inline fsck/repair.c:64]: I think ERP_FIX is better than ERP_TRUNCATE. because this makes user's asking "FIx?".

File: fsck/repair.h [hunk 5]
```diff
@@ -23,7 +23,7 @@
 #define ER_FILE_LARGER_SIZE		0x00002005
 #define ER_FILE_DUPLICATED_CLUS		0x00002006
 #define ER_FILE_ZERO_NOFAT		0x00002007
-
+#define ER_DIR_FIRST_CLUS		0x00002008
```
Comment (hclee) [inline fsck/repair.h:26]: Could you rename this to ER_DE_XXXX instead of ER_DIR_XXXX?

File: include/exfat_dir.h [hunk 6]
```diff
@@ -75,6 +75,7 @@ int exfat_update_file_dentry_set(struct exfat *exfat,
 int exfat_build_file_dentry_set(struct exfat *exfat, const char *name,
 				unsigned short attr, struct exfat_dentry **dentry_set,
 				int *dentry_count);
+int find_free_cluster(struct exfat *exfat, clus_t start, clus_t *new_clu);
```
Comment (hclee) [inline include/exfat_dir.h:78]: it seems better to append exfat prefix to name.

**Review by hclee — COMMENTED** [review]
There are many unnecessary commits in your PR. Can you rebase again?
And it is good if you add a test-case for this to tests directory.

File: fsck/fsck.c (hunk 1 above)
Comment (Erichorng) [reply fsck/fsck.c:230]: Sorry,  forget to remove this line.

File: fsck/fsck.c (hunk 2 above)
Comment (Erichorng) [reply fsck/fsck.c:256]: OK.

**Pull request closed without merging**